### ISSUE TITLE
Registration required on child container for proper disposal

### ISSUE THREAD

**Issue by tonysneed** [grumpydev/TinyIoC#62]
If services are registered on a parent container, calling Dispose on a child container will not call dispose on registered services.  Services must be registered directly on each child container. This is not easy to do with a Web API dependency resolver, where you would create a child container on BeginScope.  The only workaround I have found is to call AutoRegister on the child container, but this is not ideal.

It seems like calling Dispose on a child container should Dispose dependencies that were registered on the parent container if they were not also registered on the child container?

**Comment by MihaMarkic** [comment]
Has this ever been solved? @tonysneed raised a valid point and @jonie provided a valid example of how it should be. If that's not what it should be, then either we misunderstood the purpose if child container or there is another way to deal with limited lifetimes.

**Comment by niemyjski** [comment]
I don't think so, but I haven't dived in... Would you mind taking a look? I'd agree with @kentcb's last comment.

**Comment by MihaMarkic** [comment]
@niemyjski  Here is an example. I would register all stuff globally, at application start. And at some point I'd like to have a limited lifespan. Doing it your way, I'd have to create child container and register everything again (or at least what I need) even though it is already registered.
Why not create a child container, use it and then dispose it to clear whatever was created?
What purpose has child container otherwise? To override few registrations?

**Comment by niemyjski** [comment]
That makes sense and is straight forward. My concern though, is if you have a parent and child containers and you have a parent object like `IFoo` and you dispose of the child container. Then dispose should be called on all objects in the child container. So now you have a disposed object registered in the parent which might be a singleton.

**Comment by MihaMarkic** [comment]
@niemyjski Singletons shouldn't be disposed by child container IMO.

**Comment by niemyjski** [comment]
If you dispose the child container than dispose should be called on the
singletons registered in the child container..

**Comment by MihaMarkic** [comment]
@niemyjski You mean "singletons that are _resolved and created (because they don't exist on parent container)_ on child container"? That's an option as well. Either that or transfer ownership to parent-most container. I would assume the later, but both should work.

**Comment by niemyjski** [comment]
Seems weird that parent and child containers could have different scopes but share the same instance. It's almost like if anything is registered in the child container we dispose of it, otherwise get it from the parent container instance. I'm just trying to think of what's straight forward, but I'm all in for what the general consensus is.
